### DISCUSSION TITLE
Turn by Turn: Always Turn Right

### DISCUSSION THREAD

**Discussion by mertbakir** [valhalla/valhalla#3289] [Q&A] [answered]
Hi, I wonder if it's possible to set a restriction like "always turn right" or set a penalty for left turns. If possible, is it also possible to apply these restrictions for a specified area (polygon) ?

**Comment by kevinkreiser** [comment, marked as answer]
it would be pretty straight forward for us to add support for request time modification of turn penalties (we already have hardcoded ones).

doing it only inside the polygon is quite advanced and probably prohibatively costly to do it on the fly. instead we'd have to mark all the graph nodes that are inside the polygon upfront (as the first step before we run the route) and then store them in a map to check if the penalty should be applied at any point in the algorithm.

its all doable but none of it exists today. if you are interested in working on it i can show you where the changes would be made

**Reply by mertbakir** [reply]
So, for the polygon part you are saying that I should edit the osm.pbf fille. Did I understand correctly? 

Also, I would like to know where the changes would be made. Thanks

**Reply by nilsnolde** [reply]
+1 for customization of (left) turn penalties, at least for motorized vehicles.

> for the polygon part you are saying that I should edit the osm.pbf fille

I think he's referring to the way valhalla internally intersects the graph nodes with the polygon. either one could let the routing algorithm intersect nodes (prohibitively expensive) or in pre-processing step, between receiving the request and traversing the edges, to produce a set of all intersected nodes the routing algorithm can check at runtime. you'd leave the osm data the way it is. we do the same with exclude_polygons, so much of the logic is already there. 

IMO the functionality to pass polygons to restrict _other parameters_ could be really cool especially for applying or lifting hard restrictions (like disable oneways etc). that'd be a little broader design question IMO

**Reply by kevinkreiser** [reply]
yep this is what i meant, sorry for the obtuse wording :smile: but also thank you for reading my mind haha

**Comment by mertbakir** [comment]
I didn't open a new discussion because the question is similar. Is there a way to avoid u-turns?

Also, can we easily implement the left turn penalties? The polygon idea seems more complicated (but would be cool in my opinion), still applying the same penalty for whole route is as easy as adding/changing some multipliers if I understand correctly. Should I open another discussion as a feature request?

**Reply by kevinkreiser** [reply]
allowing more user control in penalties for maneuver type would work the same as it would for other types (left, right, etc). if that is sufficient scope for what you are looking to do, then i think we could create an issue where we discuss the implementation and then see who has the time to implement it as a feature request. and yes it should be as adding new request parameters that we can use as multipliers on existing penalties or perhaps as absolute penalties we add  to the existing ones. we'll ahve to see what  is the best approach. the implementation is quick but we will need to do some fairly involved QA to set  sane limits on these values.

**Reply by mertbakir** [reply]
This scope is very much sufficient for me. I just opened an issue: https://github.com/valhalla/valhalla/issues/3296

**Reply by mertbakir** [reply]
@kevinkreiser Hi again, can you point me the code part that needs to be modified for this subject. I don't know C but I feel like I can try in a weekend or something. Also, do you know another alternative routing engine or a paid service that covers this feature?

Another question, do you think similar result can be achieved by manipulating the source map data that valhalla is using?
